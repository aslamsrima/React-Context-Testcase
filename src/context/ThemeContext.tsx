import { Dispatch, SetStateAction, createContext, useContext } from 'react';

export type Theme = 'light' | 'dark';

interface ThemeContextProps {
  theme: Theme;
  toggleTheme: () => void;
  setTheme: Dispatch<SetStateAction<Theme>>;
}

const ThemeContext = createContext<ThemeContextProps | undefined>(undefined);

const useTheme = (): ThemeContextProps => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};

export { ThemeContext, useTheme };
