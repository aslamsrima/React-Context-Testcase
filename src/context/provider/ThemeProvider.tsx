import React, { useState, ReactNode } from 'react';
import { Theme, ThemeContext } from '../ThemeContext';

const initialTheme: Theme = 'light';

interface ThemeProviderProps {
  children: ReactNode;
}

const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>(initialTheme);

  const toggleTheme = () => {
    setTheme((prevTheme) => (prevTheme === 'light' ? 'dark' : 'light'));
  };

  return <ThemeContext.Provider value={{ theme, toggleTheme, setTheme }}>{children}</ThemeContext.Provider>;
};

export default ThemeProvider;
