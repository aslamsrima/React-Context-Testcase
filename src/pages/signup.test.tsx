import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useTheme } from '../context/ThemeContext';
import ThemeProvider from '../context/provider/ThemeProvider';
import { act } from 'react-dom/test-utils';
import SignUp from './signup';

jest.mock('../context/ThemeContext', () => ({
  ...jest.requireActual('../context/ThemeContext'),
  useTheme: jest.fn(),
}));

jest.mock('../services/mockAPIs', () => ({
  mockSignUp: jest.fn(),
}));

describe('Login Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot test', async () => {
    (useTheme as jest.Mock).mockReturnValue({ theme: 'light', setTheme: jest.fn(), toggleTheme: jest.fn() });

    const app = render(
      <ThemeProvider>
        <SignUp />
      </ThemeProvider>,
    );
    expect(app).toMatchSnapshot();
  });

  test('Login component renders and submits form with validation error', async () => {
    const app = render(
      <ThemeProvider>
        <SignUp />
      </ThemeProvider>,
    );
    fireEvent.click(app.getByText('Login'));

    await waitFor(() => {
      expect(require('../services/mockAPIs').mockSignUp).toHaveBeenCalledTimes(0);
    });
  });

  test('Login component renders and submits form with mock API call', async () => {
    const app = render(
      <ThemeProvider>
        <SignUp />
      </ThemeProvider>,
    );
    act(() => {
      userEvent.type(app.getByPlaceholderText('John Doe'), 'tom latham');
      userEvent.type(app.getByPlaceholderText('info@mailaddress.com'), 'test@example.com');
      userEvent.type(app.getByPlaceholderText('••••••••••••'), 'password123');
    });
    fireEvent.click(app.getByText('Register'));

    await waitFor(() => {
      expect(require('../services/mockAPIs').mockSignUp).toHaveBeenCalledWith({
        name: 'tom latham',
        email: 'test@example.com',
        password: 'password123',
      });
      expect(require('../services/mockAPIs').mockSignUp).toHaveBeenCalledTimes(1);
    });
  });
});
