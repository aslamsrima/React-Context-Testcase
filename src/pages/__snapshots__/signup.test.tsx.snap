// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="align"
      >
        <div
          class="grid alignItem"
        >
          <div
            class="register"
          >
            <svg
              class="siteLogo"
              height="84"
              viewBox="77.7 214.9 274.7 412"
              width="56"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <lineargradient
                  id="a"
                  x1="0%"
                  y1="0%"
                  y2="0%"
                >
                  <stop
                    offset="0%"
                    stop-color="#8ceabb"
                  />
                  <stop
                    offset="100%"
                    stop-color="#378f7b"
                  />
                </lineargradient>
              </defs>
              <path
                d="M215 214.9c-83.6 123.5-137.3 200.8-137.3 275.9 0 75.2 61.4 136.1 137.3 136.1s137.3-60.9 137.3-136.1c0-75.1-53.7-152.4-137.3-275.9z"
                fill="url(#a)"
              />
            </svg>
            <h2>
              Sign Up
            </h2>
            <form
              action=""
              class="authForm"
              method="post"
            >
              <div
                class="formField"
              >
                <input
                  name="name"
                  placeholder="John Doe"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="formField"
              >
                <input
                  name="email"
                  placeholder="info@mailaddress.com"
                  type="email"
                  value=""
                />
              </div>
              <div
                class="formField"
              >
                <input
                  name="password"
                  placeholder="••••••••••••"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="formField"
              >
                <input
                  class="cursorPointer"
                  type="submit"
                  value="Register"
                />
              </div>
            </form>
            <p>
              Already have an account? 
              <a
                href="/login"
              >
                Login
              </a>
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="align"
    >
      <div
        class="grid alignItem"
      >
        <div
          class="register"
        >
          <svg
            class="siteLogo"
            height="84"
            viewBox="77.7 214.9 274.7 412"
            width="56"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <lineargradient
                id="a"
                x1="0%"
                y1="0%"
                y2="0%"
              >
                <stop
                  offset="0%"
                  stop-color="#8ceabb"
                />
                <stop
                  offset="100%"
                  stop-color="#378f7b"
                />
              </lineargradient>
            </defs>
            <path
              d="M215 214.9c-83.6 123.5-137.3 200.8-137.3 275.9 0 75.2 61.4 136.1 137.3 136.1s137.3-60.9 137.3-136.1c0-75.1-53.7-152.4-137.3-275.9z"
              fill="url(#a)"
            />
          </svg>
          <h2>
            Sign Up
          </h2>
          <form
            action=""
            class="authForm"
            method="post"
          >
            <div
              class="formField"
            >
              <input
                name="name"
                placeholder="John Doe"
                type="text"
                value=""
              />
            </div>
            <div
              class="formField"
            >
              <input
                name="email"
                placeholder="info@mailaddress.com"
                type="email"
                value=""
              />
            </div>
            <div
              class="formField"
            >
              <input
                name="password"
                placeholder="••••••••••••"
                type="password"
                value=""
              />
            </div>
            <div
              class="formField"
            >
              <input
                class="cursorPointer"
                type="submit"
                value="Register"
              />
            </div>
          </form>
          <p>
            Already have an account? 
            <a
              href="/login"
            >
              Login
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
